File: code-generator/src/main/java/io/github/potjerodekool/openapi/internal/di/ClassPathScanner.java
package io.github.potjerodekool.openapi.internal.di;

import io.github.potjerodekool.openapi.internal.di.bean.BeanDefinition;
import org.objectweb.asm.ClassReader;

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.PathMatcher;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ClassPathScanner {

    private static final PathMatcher CLASS_FILE_PATH_MATCHER = FileSystems.getDefault().getPathMatcher("glob:**.class");

    private ClassPathScanner() {
    }

    public static List<BeanDefinition> scan() {
        final var beanDefinitions = new ArrayList<BeanDefinition>();
        try {
            final var resources = ClassPathScanner.getClassLoader()
                    .getResources("META-INF/io.github.potjerodekool.openapi.common.autoconfig.AutoConfiguration");

            final var iterator = resources.asIterator();

            while (iterator.hasNext()) {
                final var resource = iterator.next();

                try (final var inputStream = resource.openStream()) {
                    final var lines = new String(inputStream.readAllBytes()).split("\n");

                    Arrays.stream(lines)
                            .filter(line -> !line.trim().startsWith("#"))
                            .forEach(className -> {
                                beanDefinitions.addAll(loadConfiguration(className));
                            });
                }
            }

        } catch (final IOException e) {
            throw new RuntimeException(e);
        }

        return beanDefinitions;
    }

    private static List<BeanDefinition> loadConfiguration(final String className) {
        try (final var inputStream = getClassLoader().getResourceAsStream(
                className.replace('.', '/') + ".class"
        )) {
            final var instance = getClassLoader().loadClass(className).getDeclaredConstructor().newInstance();
            final var data = inputStream.readAllBytes();
            final var classReader = new ClassReader(data);
            final var reader = new AutoConfigReader(instance);
            classReader.accept(reader, 0);
            return reader.getBeanDefinitions();
        } catch (final Exception e) {
            //Ignore
            e.printStackTrace();
        }
        return List.of();
    }

    protected static ClassLoader getClassLoader() {
        final var classLoader = ClassPathScanner.class.getClassLoader();

        if (classLoader != null) {
            return classLoader;
        } else {
            return ClassLoader.getSystemClassLoader();
        }
    }
}

File: code-generator/src/main/kotlin/io/github/potjerodekool/openapi/internal/di/ClassPathScanner.kt
package io.github.potjerodekool.openapi.internal.di

import io.github.potjerodekool.openapi.internal.di.bean.BeanDefinition
import org.objectweb.asm.ClassReader
import java.io.IOException
import java.util.*

object ClassPathScanner {
    fun scan(): List<BeanDefinition> {
        val beanDefinitions = ArrayList<BeanDefinition>()
        try {
            val resources = classLoader
                .getResources("META-INF/io.github.potjerodekool.openapi.common.autoconfig.AutoConfiguration")

            val iterator = resources.asIterator()

            while (iterator.hasNext()) {
                val resource = iterator.next()

                resource.openStream().use { inputStream ->
                    val lines = String(inputStream.readAllBytes()).split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray()
                    Arrays.stream(lines)
                        .filter { line: String -> !line.trim { it <= ' ' }.startsWith("#") }
                        .forEach { className: String -> beanDefinitions.addAll(loadConfiguration(className)) }
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return beanDefinitions
    }

    private fun loadConfiguration(className: String): List<BeanDefinition> {
        try {
            classLoader.getResourceAsStream(
                className.replace('.', '/') + ".class"
            ).use { inputStream ->
                if (inputStream == null) {
                    return listOf()
                }
                val data = inputStream.readAllBytes()
                val instance = classLoader.loadClass(className).getDeclaredConstructor().newInstance()
                val classReader = ClassReader(data)
                val reader = AutoConfigReader(instance)
                classReader.accept(reader, 0)
                return reader.beanDefinitions
            }
        } catch (e: Exception) {
            //Ignore
            e.printStackTrace();
        }
        return listOf()
    }

    val classLoader: ClassLoader
        get() {
            val classLoader = ClassPathScanner::class.java.classLoader

            return classLoader ?: ClassLoader.getSystemClassLoader()
        }
}
